src/jit/lowerxarch.cpp

```diff
@@ -3763,6 +3867,63 @@ bool Lowering:: IsContainableImmed(GenTree* parentNode, GenTree* childNode)
     return true;
 }
 
+//----------------------------------------------------------------------
+// SetRegOptional - sets a bit to indicate to LSRA that register for
+// for a given tree node is optional for codegen purpose.
+//
+// Arguments:
+//    tree    -   GenTree node
+//
+// Returns
+//    None
+//
+// Note: Right now a tree node is marked as reg optional only
+// if is it a GT_LCL_VAR.  This routine needs to be modified if
+// in future if lower/codegen needs to support other tree node
+// types.
+void Lowering::SetRegOptional(GenTree* tree)
+{
+    if (tree->OperGet() == GT_LCL_VAR)
+    {
+        tree->gtLsraInfo.regOptional = true;
+    }
+}
+
+// ------------------------------------------------------------------
+// SetRegOptionalBinOp - Indicates which of the operands of a bin-op
+// register requirement is optional. Xarch instruction set allows
+// either of op1 or op2 of binary operation (e.g. add, mul etc) to be
+// a memory operand.  This routine provides info to register allocator
+// which of its operands optionally require a register.  Lsra might not
+// allocate a register to RefTypeUse positions of such operands if it
+// is beneficial. In such a case codegen will treat them as memory
+// operands.
+//
+// Arguments:
+//     tree  -  Gentree of a bininary operation.
+//
+// Returns 
+//     None.
+void Lowering::SetRegOptionalForBinOp(GenTree* tree)
+{
+    assert(GenTree::OperIsBinary(tree->OperGet()));
+
+    GenTree* op1 = tree->gtGetOp1();
+    GenTree* op2 = tree->gtGetOp2();
+
+    if (tree->TypeGet() == op2->TypeGet())
+    {
+        SetRegOptional(op2);
+    }
+
```

Comment:
I don't think that your example is applicable here.
both op1 and op2 can't get marked as SetRegOptional.

If you successfully marked op2 as RegOptional, then by the very first check in the next if statement you won't try to mark op1 as RegOptional.  Using an else clause here would just avoid doing that check.